omod/src/main/java/org/openmrs/module/referencedemodata/web/controller/DemoDataGenerationController.java

```diff
@@ -0,0 +1,96 @@
+package org.openmrs.module.referencedemodata.web.controller;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.openmrs.GlobalProperty;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.referencedemodata.ReferenceDemoDataActivator;
+import org.openmrs.module.referencedemodata.ReferenceDemoDataConstants;
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.v1_0.controller.BaseRestController;
+import org.openmrs.scheduler.TaskDefinition;
+import org.openmrs.scheduler.tasks.AbstractTask;
+import org.openmrs.util.PrivilegeConstants;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/" + ReferenceDemoDataConstants.REFERENCE_DEMO_DATA_URI)
+public class DemoDataGenerationController extends BaseRestController {
+	
+	private static ReferenceDemoDataActivator referenceDemoDataActivator;
+	
+	private static final String REFERENCE_DEMO_DATA_TASK_NAME = "Reference demo data generation task";
+	
+	public static final String NUMBER_OF_DEMO_PATIENTS_PARAMETER = "numberOfDemoPatients";
+
+	@RequestMapping(value = ReferenceDemoDataConstants.GENERATE_DEMO_DATA_URI, method = RequestMethod.GET, produces = "application/json")
+	@ResponseBody()
+	public String generateDemoData(HttpServletRequest request) throws Exception {
+		
+		String numberOfDemoPatients = request.getParameter(NUMBER_OF_DEMO_PATIENTS_PARAMETER);
+		
+		try {
+			Integer.parseInt(numberOfDemoPatients);
+		}
+		catch (NumberFormatException e) {
+			String errorMsg = String.format("Could not parse '%s' as an integer", numberOfDemoPatients);
+			return String.format("{\"error\": \"%s\"}", errorMsg);
+		}
+		
+		Context.addProxyPrivilege(PrivilegeConstants.MANAGE_SCHEDULER);
+		TaskDefinition taskDef = Context.getSchedulerService().getTaskByName(REFERENCE_DEMO_DATA_TASK_NAME);
+		if (taskDef == null) {
+			taskDef = new TaskDefinition();
+			taskDef.setTaskClass(GenerateDemoDataTask.class.getName());
+			taskDef.setStartOnStartup(false);
+			taskDef.setStartTime(null); // Executes immediately
+			taskDef.setName(REFERENCE_DEMO_DATA_TASK_NAME);
+			taskDef.setDescription(REFERENCE_DEMO_DATA_TASK_NAME);
+			taskDef.setRepeatInterval(0L); // Executes once
+			taskDef.setProperty("numberOfDemoPatients", numberOfDemoPatients);
+			Context.getSchedulerService().saveTaskDefinition(taskDef);
+		}
+		if (Context.getPatientService().getAllPatients().size() < Integer.parseInt(numberOfDemoPatients)) {
```

Comment:
So this treats `numberOfDemoPatients` as though it's the goal number for # of patients, i.e., this is now a request to "ensure the system has at least `x` patients", which seems (to me) counter-intuitive. I would just treat this parameter for the number of patients to generate. (This is also how the Activator already works, i.e., if you set the GP to 50, it will generate 50 patients regardless of how many exist).

Reply:
Thinking through, this seemed somewhat more safe since there is nothing that can act as a `Are you sure you want to generate x patients` button. This can be a nice check so anyone doing this doesn't do it by mistake. Also if this trigger is always happening at start and re-start by an external automated request, the trigger would un-necessary create more patients, which otherwise can be done in two calls (one querying total number and another triggering generation if necessary).

Reply:
Sure, but we're talking about a request that should be limited to a select number of users anyways (this method really should have some kind of access control, i.e., requiring that the user has `MANAGE_GLOBAL_PROPERTIES` or something would be good...

If we want to support this "maximum number of patients" behaviour as a use case, I think it makes more sense to make it opt-in than the default behaviour, e.g., if the request is sent with some header or an additional parameter, etc.

Reply:
Thanks for the review @ibacher, makes sense. I've applied the ideas in the latest commit.

Reply:
@ibacher could you review again thanks.